sdk/core/azure-core/test/ut/http_method.cpp

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <gtest/gtest.h>
+
+#include <azure/core/http/http.hpp>
+
+using namespace Azure::Core::Http;
+
+TEST(HttpMethod, Get)
+{
+  HttpMethod const get = HttpMethod::Get;
+  EXPECT_EQ(get.ToString(), "GET");
+
+  EXPECT_EQ(HttpMethod::Get, HttpMethod::Get);
+  EXPECT_EQ(get, get);
+  EXPECT_EQ(get, HttpMethod("GET"));
+  EXPECT_EQ(HttpMethod("GET"), HttpMethod::Get);
+
+  EXPECT_NE(get, HttpMethod("Get"));
+  EXPECT_NE(HttpMethod("Get"), get);
+
+  EXPECT_NE(get, HttpMethod(std::string()));
+  EXPECT_NE(HttpMethod(std::string()), get);
+
+  EXPECT_EQ(get, HttpMethod::Get);
+  EXPECT_EQ(HttpMethod::Get, get);
+
+  EXPECT_NE(get, HttpMethod::Head);
+  EXPECT_NE(get, HttpMethod::Post);
+  EXPECT_NE(get, HttpMethod::Put);
+  EXPECT_NE(get, HttpMethod::Delete);
+  EXPECT_NE(get, HttpMethod::Patch);
+  EXPECT_NE(get, HttpMethod("TRACE"));
+  EXPECT_NE(get, HttpMethod("HUMUHUMUNUKUNUKUAPUAA"));
+
+  EXPECT_NE(HttpMethod::Head, get);
+  EXPECT_NE(HttpMethod::Post, get);
+  EXPECT_NE(HttpMethod::Put, get);
+  EXPECT_NE(HttpMethod::Delete, get);
+  EXPECT_NE(HttpMethod::Patch, get);
+  EXPECT_NE(HttpMethod("TRACE"), get);
+  EXPECT_NE(HttpMethod("HUMUHUMUNUKUNUKUAPUAA"), get);
+}
+
+TEST(HttpMethod, Head)
+{
+  HttpMethod const head = HttpMethod::Head;
+  EXPECT_EQ(head.ToString(), "HEAD");
+
+  EXPECT_EQ(HttpMethod::Head, HttpMethod::Head);
+  EXPECT_EQ(head, head);
+  EXPECT_EQ(head, HttpMethod("HEAD"));
+  EXPECT_EQ(HttpMethod("HEAD"), HttpMethod::Head);
+
+  EXPECT_NE(head, HttpMethod("Head"));
+  EXPECT_NE(HttpMethod("Head"), head);
+
+  EXPECT_NE(head, HttpMethod(std::string()));
+  EXPECT_NE(HttpMethod(std::string()), head);
+
+  EXPECT_EQ(head, HttpMethod::Head);
+  EXPECT_EQ(HttpMethod::Head, head);
+
+  EXPECT_NE(head, HttpMethod::Get);
+  EXPECT_NE(head, HttpMethod::Post);
+  EXPECT_NE(head, HttpMethod::Put);
+  EXPECT_NE(head, HttpMethod::Delete);
+  EXPECT_NE(head, HttpMethod::Patch);
+  EXPECT_NE(head, HttpMethod("TRACE"));
+  EXPECT_NE(head, HttpMethod("HUMUHUMUNUKUNUKUAPUAA"));
+
+  EXPECT_NE(HttpMethod::Get, head);
+  EXPECT_NE(HttpMethod::Post, head);
+  EXPECT_NE(HttpMethod::Put, head);
+  EXPECT_NE(HttpMethod::Delete, head);
+  EXPECT_NE(HttpMethod::Patch, head);
+  EXPECT_NE(HttpMethod("TRACE"), head);
+  EXPECT_NE(HttpMethod("HUMUHUMUNUKUNUKUAPUAA"), head);
```

Comment:
A lot of this is being duplicated for each of the enum values. Could we extract these to a helper function instead to avoid duplication?

Reply:
It can be done, but the failure logs aren't probably going to be that clear, and the code would not be identical, because it would not use `HttpMethod::<Method>`, but a copy of it instead, which is technically not the same thing, and likely do not reflect the most common usage in user's code.

Reply:
So, I would rather leave it as is. Certainly would've been easier to remove the dup - it took me some time to have almost identical code repeated correctly. But that was kind of an intent - to use primitive methods and make minimum assumptions, and just write plain code that does verification. But I am not super strong on this one. If you feel strongly, I can refactor it into one method that gets invoked with different parameters.